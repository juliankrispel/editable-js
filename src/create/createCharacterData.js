// @flow

import type { RawBlock, CharacterData } from '../types'

export default function createCharacterData(block: RawBlock): Array<CharacterData> {
  if (block.text != null) {
    return block.text.split('').map((_, index) => {
      const data = {
        styles: [],
        entities: []
      }

      if (block.characterRanges != null) {
        block.characterRanges.filter(range => {
          return range.start <= index &&
          range.offset + range.start > index
        }).forEach(range => {
          if (Array.isArray(range.styles)) {
            data.styles = data.styles.concat(range.styles)
          }

          if (Array.isArray(range.entities)) {
            data.entities = data.entities.concat(range.entities)
          }
        })
      }

      return data
    })
  }
  return []
}
