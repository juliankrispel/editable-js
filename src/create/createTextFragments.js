// @flow
import type { TextFragment, Block, EntityMap } from '../types'

const hasEqualMarks = (left, right) =>
  right.entity === left.entity &&
  Array.from(left.marks).sort().join('') === Array.from(right.marks).sort().join('')

export default function createTextFragments(block: Block, entityMap: EntityMap): Array<TextFragment> {
  return block.characterData.reduce(
    (acc, data, index) => {
      if (acc.length < 1) {
        return [{
          marks: data.marks,
          entity: data.entity,
          offset: index,
          text: block.text[index]
        }]
      } else {
        const lastFragment = acc[acc.length - 1]
        if (hasEqualMarks(lastFragment, data)) {
          return acc.slice(0, -1).concat([{
            ...lastFragment,
            text: lastFragment.text + block.text[index]
          }])
        } else {
          return [
            ...acc,
            {
              marks: data.marks,
              entity: data.entity,
              offset: index,
              text: block.text[index]
            }
          ]
        }
      }
    },
    []
  ).map(fragment => ({
    ...fragment,
    entity: fragment.entity != null ? entityMap[fragment.entity] : null,
  }))
}
