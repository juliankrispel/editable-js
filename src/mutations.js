// @flow

import type { RawBlock, EditorState, SelectionState, Block } from './types'
import genId from './genId'

export const createBlock = (newBlock: RawBlock): Block => ({
  ...newBlock,
  key: genId()
})

export const insertText = (
  editorState: EditorState,
  text: string
): void => {
  const { selection } = editorState

  if (!isCollapsed(selection)) {
    throw new Error('cannot insert text when selection is not collapsed')
  }

  const { startOffset, startKey } = selection

  editorState.content.forEach(block => {
    if (block.key === startKey && typeof block.value === 'string') {
      const { value } = block
      block.value = `${value.slice(0, startOffset)}${text}${value.slice(startOffset)}`
    }
  })

  const offset = selection.startOffset + text.length

  editorState.selection.startOffset = offset
  editorState.selection.endOffset = offset
}

export const replaceText = (editorState: EditorState, selection: SelectionState, text: string): void => {
  removeRange(editorState, editorState.selection)
  insertText(editorState, text)
}

export const removeRange = (editorState: EditorState, selection: SelectionState): void => {
  const { content } = editorState
  const { startKey, endKey, startOffset, endOffset } = selection
  const keys = content.map(block => block.key)
  const startIndex = keys.indexOf(startKey)
  const endIndex = keys.indexOf(endKey)

  console.log('yo, selection', selection)

  if (selection.startKey === selection.endKey) {
    const { value } = content[startIndex]
    content[startIndex].value = value.slice(0, startOffset) + value.slice(endOffset)
  } else {
    content[startIndex].value = content[startIndex].value.slice(0, startOffset) + content[endIndex].value.slice(endOffset)
    content.splice(endIndex, 1)
  }

  content.forEach((block, index) => {
    if (!(index <= startIndex || endIndex <= index)) {
      content.splice(index, 1)
    }
  })

  editorState.selection = collapseToStart(selection)
}

export const collapseToStart = (selection: SelectionState): SelectionState => ({
  ...selection,
  endOffset: selection.startOffset,
  endKey: selection.startKey
})

export const splitBlock = (editorState: EditorState, selection: SelectionState): void => {
  if (!isCollapsed(selection)) {
    removeRange(editorState, selection)
  }

  const { startOffset, endOffset, startKey } = selection

  const blockToSplit = getBlockFor(editorState, startKey)

  if (blockToSplit == null) {
    return
  }

  const textBefore = blockToSplit.value.slice(0, startOffset)
  const textAfter = blockToSplit.value.slice(endOffset)

  const newBlock = createBlock({ ...blockToSplit, value: textAfter })

  updateBlock(editorState, blockToSplit.key, { value: textBefore })
  insertBlockAfter(editorState, newBlock)

  editorState.selection = {
    startKey: newBlock.key,
    endKey: newBlock.key,
    startOffset: 0,
    endOffset: 0
  }
}

export const updateBlock = (editorState: EditorState, key: string, blockUpdate: $Shape<Block>): void => {
  const keys = editorState.content.map(block => block.key)
  const blockIndex = keys.indexOf(key)

  editorState.content[blockIndex] = Object.assign(
    editorState.content[blockIndex],
    blockUpdate
  )
}

export const insertBlockBefore = (editorState: EditorState, block: Block): void => {
  const { content, selection } = editorState
  const keys = content.map(block => block.key)
  const blockIndex = keys.indexOf(selection.endKey)

  editorState.content = content.slice(0, blockIndex)
    .concat([block]).concat(content.slice(blockIndex))
}

export const insertBlockAfter = (editorState: EditorState, block: Block): void => {
  const { content, selection } = editorState
  const keys = content.map(block => block.key)
  const blockIndex = keys.indexOf(selection.endKey)

  editorState.content = content.slice(0, blockIndex + 1)
    .concat([block]).concat(content.slice(blockIndex + 1))
}

export const isCollapsed = (selection: SelectionState): boolean =>
  selection.startOffset === selection.endOffset &&
  selection.startKey === selection.endKey

export const getBlockFor = ({ content }: EditorState, key: string): ?Block =>
  content.find(block => block.key === key)

export const getBlockBefore = ({ content }: EditorState, key: string): ?Block =>
  content[content.map(block => block.key).indexOf(key) - 1]

export const getBlockAfter = ({ content }: EditorState, key: string): ?Block =>
  content[content.map(block => block.key).indexOf(key) + 1]
