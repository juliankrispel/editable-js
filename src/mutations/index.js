// @flow

import collapseBlock from './collapseBlock'
import deleteBlock from './deleteBlock'
import insertBlockAfter from './insertBlockAfter'
import insertBlockBefore from './insertBlockBefore'
import insertText from './insertText'
import replaceText from './replaceText'
import mergeBlock from './mergeBlock'
import removeRange from './removeRange'
import splitBlock from './splitBlock'
import updateBlock from './updateBlock'

export {
  collapseBlock,
  deleteBlock,
  insertBlockAfter,
  insertBlockBefore,
  insertText,
  replaceText,
  mergeBlock,
  removeRange,
  splitBlock,
  updateBlock
}
