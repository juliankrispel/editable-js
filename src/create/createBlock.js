// @flow

import type { RawBlock, Block } from '../types'
import createId from './createId'
import createCharacterData from './createCharacterData'

export default function createBlock (block: RawBlock): Block {
  const { children, text = '', key, characterRanges, ...rest } = block

  const newBlock = {
    ...rest,
    text,
    characterData: createCharacterData(block),
    children: [],
    key: key || createId()
  }

  if (Array.isArray(children)) {
    newBlock.children = children.map(createBlock)
  }

  return newBlock
}
