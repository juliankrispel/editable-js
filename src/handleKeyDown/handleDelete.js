// @flow

import { getBlockAfter, getBlock } from '../queries'
import { isCollapsed } from '../selection'
import { removeRange } from '../mutations'

import type { EditorState } from '../types'

export default function handleDelete (editorState: EditorState) {
  if (!isCollapsed(editorState.selection)) {
    removeRange(editorState, editorState.selection)
  }

  const { content, selection } = editorState
  const blockAfter = getBlockAfter(content, selection.endKey)
  const currentBlock = getBlock(content, selection.endKey)

  if (currentBlock == null) {
    throw new Error('current block not defined')
  }

  if (selection.endOffset < currentBlock.text.length) {
    selection.endOffset = selection.endOffset
  } else if (blockAfter != null) {
    selection.endKey = blockAfter.key
    selection.endOffset = 0
  } else {
    return
  }

  removeRange(editorState, selection)
}
