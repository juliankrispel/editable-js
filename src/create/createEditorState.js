// @flow

import produce from 'immer'
import type { RawContentState, EditorState } from '../types'
import createBlock from './createBlock'

const emptyState = {
  content: [],
  entityMap: {},
  selection: {
    startKey: 'a',
    endKey: 'a',
    startOffset: 0,
    endOffset: 0
  },
  changes: [],
  changeIndex: -1,
  lastCommitted: null
}

const createEditorState = (rawContent: RawContentState): EditorState => {
  const content = rawContent.blocks.map(createBlock)

  const key = content[0].key
  const selection = {
    startKey: key,
    endKey: key,
    startOffset: 0,
    endOffset: 0
  }

  const initialState = produce(emptyState,
    () => ({
      content,
      entityMap: rawContent.entityMap,
      selection,
      changes: [],
      changeIndex: 0,
      lastCommitted: null
    }))

  return initialState
}

export default createEditorState
