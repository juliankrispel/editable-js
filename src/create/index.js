import createBlock from './createBlock'
import createEditorState from './createEditorState'

export {
  createBlock,
  createEditorState
}
