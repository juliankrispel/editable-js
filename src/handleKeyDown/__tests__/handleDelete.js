// @flow

import { createEditorState } from '../../create'
import { commit } from '../../history'
import handleDelete from '../handleDelete'
import { updateSelection } from '../../mutations'

describe('handleDelete', () => {
  const initialState = createEditorState({ blocks: [{
    text: 'Line One',
    key: '1'
  }, {
    text: 'Line Two',
    key: '2'
  }]})

  describe('when selection is not collapsed', () => {
    let newEditorState = commit(initialState, updateSelection, {
      startOffset: 4,
      startKey: '1',
      endOffset: 7,
      endKey: '2'
    })

    newEditorState = commit(newEditorState, handleDelete)

    test('removes range', () => {
      expect(newEditorState.content).toMatchSnapshot()
    })

    test('collapses selection to start', () => {
      expect(newEditorState.selection).toMatchSnapshot()
    })
  })

  describe('when selection is collapsed', () => {
    let newEditorState = commit(initialState, updateSelection, {
      startOffset: 4,
      startKey: '1',
      endOffset: 4,
      endKey: '1'
    })

    newEditorState = commit(newEditorState, handleDelete)

    test('removes next character', () => {
      expect(newEditorState.content).toMatchSnapshot()
    })

    test('selection remains the same', () => {
      expect(newEditorState.selection).toMatchSnapshot()
    })
  })

  describe('when selection is collapsed and at end of block', () => {
    let newEditorState = commit(initialState, updateSelection, {
      startOffset: initialState.content[0].text.length,
      startKey: '1',
      endOffset: initialState.content[0].text.length,
      endKey: '1'
    })

    newEditorState = commit(newEditorState, handleDelete)

    test('merges with next block', () => {
      expect(newEditorState.content).toMatchSnapshot()
    })

    test('selection remains the same', () => {
      expect(newEditorState.selection).toMatchSnapshot()
    })
  })

  describe('when selection is collapsed and at end of the document', () => {
    let newEditorState = commit(initialState, updateSelection, {
      startOffset: initialState.content[0].text.length,
      startKey: '2',
      endOffset: initialState.content[0].text.length,
      endKey: '2'
    })

    newEditorState = commit(newEditorState, handleDelete)

    test('editorState.content remains the same', () => {
      expect(newEditorState.content).toMatchSnapshot()
    })

    test('editorState.selection remains the same', () => {
      expect(newEditorState.selection).toMatchSnapshot()
    })
  })
})
