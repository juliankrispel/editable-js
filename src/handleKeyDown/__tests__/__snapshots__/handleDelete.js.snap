// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleDelete when selection is collapsed and at end of block merges with next block 1`] = `
Array [
  Object {
    "characterData": Array [
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
    ],
    "children": Array [],
    "key": "1",
    "text": "Line OneLine Two",
  },
]
`;

exports[`handleDelete when selection is collapsed and at end of block selection remains the same 1`] = `
Object {
  "endKey": "1",
  "endOffset": 8,
  "startKey": "1",
  "startOffset": 8,
}
`;

exports[`handleDelete when selection is collapsed and at end of the document editorState.content remains the same 1`] = `
Array [
  Object {
    "characterData": Array [
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
    ],
    "children": Array [],
    "key": "1",
    "text": "Line One",
  },
  Object {
    "characterData": Array [
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
    ],
    "children": Array [],
    "key": "2",
    "text": "Line Two",
  },
]
`;

exports[`handleDelete when selection is collapsed and at end of the document editorState.selection remains the same 1`] = `
Object {
  "endKey": "2",
  "endOffset": 8,
  "startKey": "2",
  "startOffset": 8,
}
`;

exports[`handleDelete when selection is collapsed removes next character 1`] = `
Array [
  Object {
    "characterData": Array [
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
    ],
    "children": Array [],
    "key": "1",
    "text": "LineOne",
  },
  Object {
    "characterData": Array [
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
    ],
    "children": Array [],
    "key": "2",
    "text": "Line Two",
  },
]
`;

exports[`handleDelete when selection is collapsed selection remains the same 1`] = `
Object {
  "endKey": "1",
  "endOffset": 4,
  "startKey": "1",
  "startOffset": 4,
}
`;

exports[`handleDelete when selection is not collapsed collapses selection to start 1`] = `
Object {
  "endKey": "1",
  "endOffset": 4,
  "startKey": "1",
  "startOffset": 4,
}
`;

exports[`handleDelete when selection is not collapsed removes range 1`] = `
Array [
  Object {
    "characterData": Array [
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
      Object {
        "marks": Array [],
      },
    ],
    "children": Array [],
    "key": "1",
    "text": "Line",
  },
]
`;
